x/ibc/04-channel/keeper/packet.go

```diff
@@ -0,0 +1,307 @@
+		)
+	default:
+		panic("invalid channel ordering type")
+	}
+
+	if !ok {
+		return nil, errors.New("failed packet verification") // TODO: sdk.Error
+	}
+
+	k.deletePacketCommitment(ctx, packet.SourcePort(), packet.SourceChannel(), packet.Sequence())
+	return packet, nil
+}
+
+// SendPacket  is called by a module in order to send an IBC packet on a channel
+// end owned by the calling module to the corresponding module on the counterparty
+// chain.
+func (k Keeper) SendPacket(ctx sdk.Context, packet exported.PacketI) error {
+	channel, found := k.GetChannel(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return errors.New("channel not found") // TODO: sdk.Error
+	}
+
+	if channel.State == types.CLOSED {
+		return errors.New("channel is closed") // TODO: sdk.Error
+	}
+
+	_, found = k.GetChannelCapability(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return errors.New("channel capability key not found") // TODO: sdk.Error
+	}
+
+	// if !AuthenticateCapabilityKey(capabilityKey) {
+	//  return errors.New("invalid capability key") // TODO: sdk.Error
+	// }
+
+	if packet.DestPort() != channel.Counterparty.PortID {
+		return errors.New("invalid packet destination port")
+	}
+
+	if packet.DestChannel() != channel.Counterparty.ChannelID {
+		return errors.New("invalid packet destination channel")
+	}
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, channel.ConnectionHops[0])
+	if !found {
+		return errors.New("connection not found") // TODO: ics03 sdk.Error
+	}
+
+	if connection.State == connectiontypes.NONE {
+		return errors.New("connection is closed") // TODO: sdk.Error
+	}
+
+	consensusState, found := k.clientKeeper.GetConsensusState(ctx, connection.ClientID)
+	if !found {
+		return errors.New("consensus state not found") // TODO: sdk.Error
+	}
+
+	if consensusState.GetHeight() >= packet.TimeoutHeight() {
+		return types.ErrPacketTimeout(k.codespace)
+	}
+
+	nextSequenceSend, found := k.GetNextSequenceSend(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return types.ErrSequenceNotFound(k.codespace, "send")
+	}
+
+	if packet.Sequence() != nextSequenceSend {
+		return types.ErrInvalidPacketSequence(k.codespace)
+	}
+
+	nextSequenceSend++
+	k.SetNextSequenceSend(ctx, packet.SourcePort(), packet.SourceChannel(), nextSequenceSend)
+	k.SetPacketCommitment(ctx, packet.SourcePort(), packet.SourceChannel(), packet.Sequence(), packet.Data()) // TODO: hash packet data
+
+	return nil
+}
+
+// RecvPacket is called by a module in order to receive & process an IBC packet
+// sent on the corresponding channel end on the counterparty chain.
+func (k Keeper) RecvPacket(
+	ctx sdk.Context,
+	packet exported.PacketI,
+	proof ics23.Proof,
+	proofHeight uint64,
+	acknowledgement []byte,
+) (exported.PacketI, error) {
+
+	channel, found := k.GetChannel(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return nil, types.ErrChannelNotFound(k.codespace, packet.SourceChannel())
+	}
```

Comment:
`recvPacket` should check the channel by `destPort` and `destPort`, refer to https://github.com/cosmos/ics/blob/master/spec/ics-004-channel-and-packet-semantics/README.md#receiving-packets